File: MockLibAndroid/SmartMockExample/src/main/java/com/crescentflare/smartmockexample/network/AuthenticationService.java
package com.crescentflare.smartmockexample.network;

import com.crescentflare.smartmockexample.data.User;

import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.POST;

/**
 * Network service: provides authentication (like logging in and out)
 */
public interface AuthenticationService
{
    /**
     * Log in
     */
    @POST("login")
    @FormUrlEncoded
    Call<User> login(@Field("username") String username, @Field("password") String password);

    /**
     * Log out
     */
    @POST("logout")
    Call<User> logout();
}

File: MockLibIOS/Example/SmartMockLib/Network/AuthenticationService.swift
//
//  AuthenticationService.swift
//  SmartMockLib Example
//
//  Provides the authentication API integration
//

import UIKit
import Alamofire
import ObjectMapper

class AuthenticationService {

    let mockableAlamofire: MockableAlamofire
    
    init(mockableAlamofire: MockableAlamofire) {
        self.mockableAlamofire = mockableAlamofire
    }
    
    func login(_ username: String, password: String, success: @escaping (User) -> Void, failure: @escaping (ApiError?) -> Void) {
        mockableAlamofire.request("login", method: .post, parameters: [ "username": username, "password": password ]).responseObject { (response: DataResponse<User>) in
            if let user = response.result.value {
                if response.response?.statusCode ?? 0 >= 200 && response.response?.statusCode ?? 0 < 300 {
                    success(user)
                    return
                }
            }
            if response.data != nil {
                failure(Mapper<ApiError>().map(JSONString: String(data: response.data!, encoding: String.Encoding.utf8) ?? "{}"))
            } else {
                failure(nil)
            }
        }
    }
    
    func logout() {
        mockableAlamofire.request("logout", method: .post).responseObject { (response: DataResponse<User>) in } // Don't care about the result in the callback
        Api.setCurrentUser(nil)
        Api.clearCookies()
    }
    
}
